File: typescript/src/NullFortyTime.spec.ts
import { NullFortyTime } from "./NullFortyTime"

describe("NullFortyTime", () => {
  describe("parse", () => {
    it("returns a new NullFortyTime", () => {
      const time = NullFortyTime.parse(10)
      expect(time).toBeInstanceOf(NullFortyTime)
    })
  })

  describe("construction", () => {
    it("sets minutes to zero", () => {
      const time = new NullFortyTime()
      expect(time.minutes).toEqual(0)
    })
  })

  describe("toString", () => {
    it("returns an empty string", () => {
      const time = new NullFortyTime()
      expect(time.toString()).toEqual("")
    })
  })

  describe("plus", () => {
    it("returns the other", () => {
      const time = new NullFortyTime()
      const other = new NullFortyTime()
      expect(time.plus(other)).toEqual(other)
    })
  })

  describe("minus", () => {
    it("returns the other", () => {
      const time = new NullFortyTime()
      const other = new NullFortyTime()
      expect(time.minus(other)).toEqual(other)
    })
  })
})

File: ruby/spec/null_forty_time_spec.rb
# frozen_string_literal: true

describe NullFortyTime do
  let(:null_forty_time) { NullFortyTime.new }

  describe 'minutes' do
    it 'returns 0' do
      expect(null_forty_time.minutes).to eq 0
    end
  end

  describe 'to_s' do
    it 'returns an empty string' do
      expect(null_forty_time.to_s).to eq ''
    end
  end

  describe 'adding with FortyTime' do
    it 'returns the other' do
      other = FortyTime.new(1)
      result = null_forty_time + other
      expect(result).to be other
      expect(result).to be_an_instance_of(FortyTime)
    end
  end

  describe 'adding with NullFortyTime' do
    it 'returns the other' do
      other = NullFortyTime.new
      result = null_forty_time + other
      expect(result).to be other
      expect(result).to be_an_instance_of(NullFortyTime)
    end
  end

  describe 'added to FortyTime' do
    it 'returns the FortyTime' do
      other = FortyTime.new(1)
      result = other + null_forty_time
      expect(result.minutes).to eq other.minutes
      expect(result).to be_an_instance_of(FortyTime)
    end
  end

  describe 'subtraction with FortyTime' do
    it 'returns the other' do
      other = FortyTime.new(1)
      result = null_forty_time - other
      expect(result).to be other
    end
  end

  describe 'subtraction with NullFortyTime' do
    it 'returns the other' do
      other = NullFortyTime
      result = null_forty_time - other
      expect(result).to be other
    end
  end

  describe 'subtracted from FortyTime' do
    it 'returns the FortyTime' do
      other = FortyTime.new(1)
      result = other - null_forty_time
      expect(result.minutes).to eq other.minutes
      expect(result).to be_an_instance_of(FortyTime)
    end
  end
end
